### PR TITLE
Add Clifford+T compiler

### PR DESCRIPTION
**Pull request by aijony**
For usage see `src/discrete/README.md`

Implements the following papers:
- Approximation of RZ as unitaries over Z[i,1/sqrt(2)] (arXiv:1212.6253v2)
- Decomposes unitaries representable by Clifford+T (arXiv:1206.5236)
- Outputs as Matsumoto Amano Normal Form (arXiv:1312.6584)

Pending on:
 - https://github.com/coalton-lang/coalton/pull/533
 - (Optional) https://github.com/coalton-lang/coalton/issues/534
 - (Optional) https://github.com/quil-lang/quilc/pull/820

### CODE REVIEW

File: src/discrete/numeric/naturals.lisp [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+  (define-class (Singleton :a)
+    (single :a))
+
+  (define-class (Singleton :a => Nat :a)
+    (fromNat (:a -> UFix)))
+
+  (define-instance (Singleton Unit)
+    (define single Unit))
+
+  (define-type Z Z)
```
Comment (eliaslfox) [inline src/discrete/numeric/naturals.lisp:17]: `(repr :enum)`

File: src/discrete/numeric/naturals.lisp [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+  (define-instance (Singleton Z)
+    (define single Z))
+
+  (define-instance (Nat Z)
+    (define (fromNat _) 0))
+
+  ;; (repr :transparent)
```
Comment (eliaslfox) [inline src/discrete/numeric/naturals.lisp:24]: Is there a reason this is commented out?

File: src/discrete/operators/ma-normal-form.lisp [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+(cl:in-package #:cl-quil.discrete/operators)
+
+(coalton-toplevel
+
+  (define-type MAForm
+    "Matsumato Amano Normal Form of Clifford+T (arXiv:0806.3834)"
+    (MAForm LeftMAForm (Backwards List MidMAForm) CliffordGates1))
+
+  (define-type LeftMAForm
```
Comment (eliaslfox) [inline src/discrete/operators/ma-normal-form.lisp:13]: `(repr :enum)`

File: src/discrete/operators/ma-normal-form.lisp [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+        ((Left-I) Nil)
+        ((Left-T) (make-list Discrete-T)))))
+
+  (define-instance (Into LeftMAForm (Unitary2 RootUnity8 (Cyclotomic8 Dyadic)))
+    (define (into c)
+      (match c
+        ((Left-I) identity)
+        ((Left-T) (into TGate)))))
+
+  (define-type MidMAForm
```
Comment (eliaslfox) [inline src/discrete/operators/ma-normal-form.lisp:35]: `(repr :enum)`

File: src/discrete/operators/rz.lisp [hunk 5]
```diff
@@ -0,0 +1,49 @@
+;;;; src/discrete/operators/rz.lisp
+;;;;
+;;;; Author: A.J. Nyquist
+
+(cl:in-package #:cl-quil.discrete/operators)
+
+(coalton-toplevel
+  (define-type (Rz :a)
```
Comment (eliaslfox) [inline src/discrete/operators/rz.lisp:8]: `(repr :transparent)`

File: src/discrete/operators/utilities.lisp [hunk 6]
```diff
@@ -0,0 +1,60 @@
+;;;; src/discrete/operators/utilities.lisp
+;;;;
+;;;; Author: A.J. Nyquist
+
+(cl:in-package #:cl-quil.discrete/operators)
+
+(coalton-toplevel
```
Comment (eliaslfox) [inline src/discrete/operators/utilities.lisp:7]: Should `Backwards` be upstreamed?

File: src/discrete/numeric/circle.lisp [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+        ((Circle a)
+         (sin (* (* 2 pi) a)))))
+    (define (circleImag a)
+      (match a
+        ((Circle a)
+         (cos (* (* 2 pi) a)))))
+    (define (circleComplex a)
+      (exp (* (* 2 pi) (complex 0 (fromCircle a))))))
+
+  (define-type RootUnity8
```
Comment (eliaslfox) [inline src/discrete/numeric/circle.lisp:52]: `(repr :transparent)`

File: src/discrete/numeric/circle.lisp [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+           (2 0)
+           (3 (Root2plex 0 (Dyadic -1 1)))
+           (4 -1)
+           (5 (Root2plex 0 (Dyadic -1 1)))
+           (6 0)
+           (7 (Root2plex 0 (Dyadic 1 1)))))
+    (define (circleComplex a)
+      (complex (circleReal a) (circleImag a))))
+
+  (define-type RootUnity4
```
Comment (eliaslfox) [inline src/discrete/numeric/circle.lisp:113]: `(repr :transparent)`

File: src/discrete/numeric/utilities.lisp [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+
+  (define-class (Monoid :a => Group :a)
+    "Type with a group-like structure (i.e. an invertable Monoid)"
+    (invert (:a -> :a)))
+
+  (define-class ((Ord :a) (Num :a) => (RealFrac :a))
+    (toRational (:a -> Fraction)))
+
+  (define-instance (RealFrac Single-Float)
+       (define (toRational x)
```
Comment (stylewarning) [inline src/discrete/numeric/utilities.lisp:17]: indentation

**Review by stylewarning — CHANGES_REQUESTED** [review]
I'm about 1/3 of the way through. I do plan on doing a couple passes and at least one complete review.

Some general comments:

- I'd like to see more toplevel documentation at the top of each file. What's the file's purpose? It's ok to be brief, as long as the reader can get a sense of what's going on.
- I'd like a README added to the src/disctete directory briefly describing the layout, what to expect to find where, and any gotchas about building.
- I'd like some documentation added to the toplevel README, especially about how to build/use/load this software. (See the section on POST_LOAD_blahblah whatever the variable is called.)

File: src/discrete/compilers/clifford-t.lisp [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+  (cl-quil:inst "H" () q))
+
+(cl-quil::define-compiler expand-Z-to-Ss
+    ((z ("Z" () q)))
+  (cl-quil:inst "S" () q)
+  (cl-quil:inst "S" () q))
+
+(cl-quil::define-compiler expand-Y-to-clifford
+    ((z ("Y" () q)))
+  ;; Y = H.S.Z.X.S.H
```
[the remark this answers is not in the document]
Comment (aijony) [reply src/discrete/compilers/clifford-t.lisp]: Whoops, I forgot to throw away phase - I'll change that to Y = Z X

File: src/discrete/discrete-common.lisp [hunk 11]
```diff
@@ -0,0 +1,35 @@
+;;;; src/discrete/discrete-chip.lisp
+;;;;
+;;;; Author: A.J. Nyquist
+
+(in-package #:cl-quil.discrete)
+
+(defparameter *deterministic-approximation* t
```
[the remark this answers is not in the document]
Comment (aijony) [reply src/discrete/discrete-common.lisp]: On second thought, I will get rid of it. The paper calls for "random" numbers for seeding, but I will just do psuedo-random.

File: app/src/entry-point.lisp [hunk 12]
```diff
@@ -34,6 +34,9 @@
     ("8Q" (quil::build-8Q-chip))
     ("20Q" (quil::build-skew-rectangular-chip 0 4 5))
     ("16QMUX" (quil::build-16QMUX-chip))
+    ("8Q-cliffordt" (funcall (alexandria::find-symbol
```
Comment (aijony) [inline app/src/entry-point.lisp:37]: Is there a better way to do this?

**Comment by ecpeterson** [comment]
Just wanted to drop by to say that this PR is _extremely_ cool for a whole spread of reasons. Really excellent work.

File: app/src/entry-point.lisp (hunk 12 above)
Comment (stylewarning) [reply app/src/entry-point.lisp:37]: We could make this into a hash table (as a part of a *separate* system) and add to the hash table elsewhere. We could leave this chip spec off in this PR and implement that in a separate PR.

**Review by stylewarning — APPROVED** [review]

**Pull request merged**